File: solutions/ocaml/forth/10/math_helpers.ml
let do_math (context : Context.t) (f : int -> int -> int) : Context.t =
  Context.with_stack context (fun stack ->
    let _ = List_helpers.require_exactly_two_values !stack in
    let x = List_helpers.shift stack in
    let y = List_helpers.shift stack in
    let result = f x y in
    stack := !stack @ [result]
  )

let add (context : Context.t) : Context.t =
     do_math context (fun x y -> x + y)

let subtract (context : Context.t) : Context.t =
     do_math context (fun x y -> x - y)

let multiply (context : Context.t) : Context.t =
     do_math context (fun x y -> x * y)

let divide (context : Context.t) : Context.t =
    do_math context (fun x y -> x / y)

File: solutions/csharp/forth/2/MathHelpers.cs
public static class MathHelpers
{
    public static Context Add(Context context)
    {
        _ = context ?? throw new ArgumentNullException(nameof(context));
        return DoMath(context, (x, y) => x + y);
    }

    public static Context Subtract(Context context)
    {
        _ = context ?? throw new ArgumentNullException(nameof(context));
        return DoMath(context, (x, y) => x - y);
    }

    public static Context Multiply(Context context)
    {
        _ = context ?? throw new ArgumentNullException(nameof(context));
        return DoMath(context, (x, y) => x * y);
    }

    public static Context Divide(Context context)
    {
        _ = context ?? throw new ArgumentNullException(nameof(context));
        return DoMath(context, (x, y) => x / y);
    }

    private static Context DoMath(Context context, Func<int, int, int> function)
        => context.WithStack(
            stack =>
            {
                int x = stack.RequireExactlyTwoValues()
                    .Shift();
                int y = stack.Shift();
                int result = function(x, y);
                stack.Add(result);
            }
        );
}

